### ISSUE TITLE
Undefined Behavior Due to Over-shift of Record Sequence Number

### ISSUE THREAD

**Issue by bathooman** [eclipse/tinydtls#57]
Hello again,

The bug occurs in the part that is responsible for checking the validity of the `record sequence number`. This bug is found by employing symbolic execution technique. When a packet is being processed (in `dtls_handle_message()`), if a peer is initialized (`dtls_peer_t *peer`), TinyDTLS checks for the validity of the `record sequence number`.  In the case that (`pkt_seq_nr > security->cseq.cseq`), we have shift operations as the [following](https://github.com/eclipse/tinydtls/blob/master/dtls.c#L3774):
```
security->cseq.bitfield <<= (pkt_seq_nr-security->cseq.cseq);  
security->cseq.bitfield |= 1<<((pkt_seq_nr-security->cseq.cseq)-1); 
```
we know that `cseq.bitfield` is an unsigned long (64-bit) and 1 is an unsigned integer (32-bit). Therefore, if the value of `pkt_seq_nr-security->cseq.cseq` is greater equal to 64 there is an over-shift for the first statement and for the values of `pkt_seq_nr-security->cseq.cseq` greater equal to 32 there is an over-shift for the second statement. Finally, since `pkt_seq_nr-security->cseq.cseq` refers to the `Record Sequence Number` in a dtls record, for the `Record Sequence Number > 32`, we will have an over-shift. 
I have attached the means to reproduce the mentioned bug. To do so, after downloading the suite, in `tinydtls-master-witness/tests`, execute the script `setupserver.sh` to compile TinyDTLS with `Undefined Behavior Sanitizor`. Note that, I have de-randomized TinyDTLS for this demonstration and the patch is available in` tinydtls-master-witness/tests/patches`. Successful execution of `setupserver.sh` will run the `dtls-server` on port 20220. Now, executing `./reproduce.sh` in `tinydtls-master-witness/tests` should result in the following run-time error from the sanitizer:

![image](https://user-images.githubusercontent.com/3464382/112139180-c33ed500-8bd2-11eb-8652-3f40cbc99dc1.png)

I hope I could explain it understandably. I appreciate the confirmation of the bug.
[tinydtls-master-witness.zip](https://github.com/eclipse/tinydtls/files/6189172/tinydtls-master-witness.zip)

Best,
Hooman Asadian

**Comment by mrdeep1** [comment]
I assume that for this case as well you are using the `feature/handshake_fragmentation` branch which is preparatory work for properly supporting fragmentation in TinyDTLS.  Issues specific to this code branch need to be fixed, as well as issues in the code in the `develop `branch.

It is appreciated your detailed analysis of the issues and providing a test environment to reproduce the issues.  These issues will be checked through and appropriate fixes raised.  However, suggested fixes are also welcome.

**Comment by bathooman** [comment]
Hello, 

These over-shift bugs also affect the Master branch. Moreover, I have provided the means to reproduce every bug I reported. If you need more information. Please let me know.

Best,
Hooman

**Comment by bathooman** [comment]
@mrdeep1 I have also checked this bug. It affects both `master` and `develop` branch.

**Comment by mrdeep1** [comment]
I agree that it is in the `develop `branch.  I am currently working on a fix.

**Comment by mrdeep1** [comment]
@bathooman Please check the fix in #60 and confirm that this fixes your issue.

**Comment by bathooman** [comment]
Dear @mrdeep1,

To the best of my knowledge and based on my tests, the pull request successfully resolves the issue.

Regards,
Hooman
